text/0000-untagged_union.md

```diff
@@ -0,0 +1,325 @@
+struct, without the requirement to mention every field of the union in a match
+or use `..`:
+
+```rust
+fn f(u: MyUnion) {
+    unsafe {
+        match u {
+            MyUnion { f1: 10 } => { println!("ten"); }
+            MyUnion { f2 } => { println!("{}", f2); }
+        }
+    }
+}
+```
+
+Matching a specific value from a union field makes a refutable pattern; naming
+a union field without matching a specific value makes an irrefutable pattern.
+Both require unsafe code.
+
+Pattern matching may match a union as a field of a larger structure.  In
+particular, when using an `untagged_union` to implement a C tagged union via
+FFI, this allows matching on the tag and the corresponding field
+simultaneously:
+
+```rust
+#[repr(u32)]
+enum Tag { I, F }
+
+untagged_union U {
+    i: i32,
+    f: f32,
+}
+
+#[repr(C)]
+struct Value {
+    tag: Tag,
+    u: U,
+}
+
+fn is_zero(v: Value) -> bool {
+    unsafe {
+        match v {
+            Value { tag: I, u: U { i: 0 } } => true,
+            Value { tag: F, u: U { f: 0.0 } } => true,
+            _ => false,
+        }
+    }
+}
+```
+
+Note that a pattern match on a union field that has a smaller size than the
+entire union must not make any assumptions about the value of the union's
+memory outside that field.
+
+## Borrowing union fields
+
+Unsafe code may borrow a reference to a field of a union; doing so borrows the
+entire union, such that any borrow conflicting with a borrow of the union
+(including a borrow of another union field or a borrow of a structure
+containing the union) will produce an error.
+
+```rust
+untagged_union U {
+    f1: u32,
+    f2: f32,
+}
+
+#[test]
+fn test() {
+    let mut u = U { f1: 1 };
+    unsafe {
+        let b1 = &mut u.f1;
+	// let b2 = &mut u.f2; // This would produce an error
+        *b1 = 5;
+    }
+    unsafe {
+        assert_eq!(u.f1, 5);
+    }
+}
+```
+
+Simultaneous borrows of multiple fields of a struct contained within a union do
+not conflict:
+
+```rust
+struct S {
+    x: u32,
+    y: u32,
+}
+
+untagged_union U {
+    s: S,
+    both: u64,
+}
+
+#[test]
+fn test() {
+    let mut u = U { s: S { x: 1, y: 2 } };
+    unsafe {
+        let bx = &mut u.s.x;
+        // let bboth = &mut u.both; // This would fail
+        let by = &mut u.s.y;
+        *bx = 5;
+        *by = 10;
+    }
+    unsafe {
+        assert_eq!(u.s.x, 5);
+        assert_eq!(u.s.y, 10);
+    }
+}
+```
+
+## Union and field visibility
+
+The `pub` keyword works on the union and on its fields, as with a struct.  The
+union and its fields default to private.  Using a private field in a union
+instantiation, field access, or pattern match produces an error.
+
+## Uninitialized unions
+
+The compiler should consider a union uninitialized if declared without an
+initializer.  However, providing a field during instantiation, or assigning to
+a field, should cause the compiler to treat the entire union as initialized.
```

Comment:
Because it seems straightforward for the compiler to detect, and unambiguously an error.  It makes sense to write one field and then pass the union to some function expecting to read that field; it never makes sense to read a field from a newly declared union that you've never written to or initialized at all.

Given your comment, I should update the RFC to clarify this paragraph to specifically reference compiler errors about accessing uninitialized variables.